x/reward/keeper/query_server.go

```diff
@@ -0,0 +1,204 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/query"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/provenance-io/provenance/x/reward/types"
+)
+
+var _ types.QueryServer = Keeper{}
+
+func (k Keeper) RewardPrograms(ctx context.Context, req *types.QueryRewardProgramsRequest) (*types.QueryRewardProgramsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	var err error
+
+	rewardProgramStates := []types.RewardProgram_State{}
+	switch req.QueryType {
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_PENDING:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_PENDING}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_ACTIVE:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_STARTED}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_FINISHED:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_FINISHED, types.RewardProgram_STATE_EXPIRED}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_OUTSTANDING:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_PENDING, types.RewardProgram_STATE_STARTED}
+	}
+
+	response := types.QueryRewardProgramsResponse{}
+	kvStore := sdkCtx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(kvStore, types.RewardProgramKeyPrefix)
+	pageResponse, err := query.FilteredPaginate(prefixStore, req.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		var rewardProgram types.RewardProgram
+		vErr := rewardProgram.Unmarshal(value)
+		switch {
+		case vErr == nil && len(rewardProgramStates) == 0:
+			if accumulate {
+				response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+			}
+		case vErr == nil:
+			for _, state := range rewardProgramStates {
+				if rewardProgram.GetState() == state {
+					if accumulate {
+						response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+					}
+					break
+				}
+			}
+		default:
+			return false, vErr
+		}
+		return true, nil
+	})
```

Comment:
This will never return `true`.

Reply:
i suck.. you're right..i consulted with @nullpointer0x00  and this is the whole function now 
```
	pageResponse, err := query.FilteredPaginate(prefixStore, req.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
		var rewardProgram types.RewardProgram
		vErr := rewardProgram.Unmarshal(value)
		switch {
		case vErr == nil && len(rewardProgramStates) == 0:
			if accumulate {
				response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
			}
			return true, nil
		case vErr == nil:
			for _, state := range rewardProgramStates {
				if rewardProgram.GetState() == state {
					if accumulate {
						response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
					}
					return true, nil
				}
			}
		default:
			return false, vErr
		}
		return false, nil
	})
```